### PR TITLE
render: Latex support

### LINKED ISSUE #135: syntax: Support latex
https://github.com/mermaid-js/mermaid/issues/1387

### PR DESCRIPTION
**Pull request by alixander**
the language tag `latex` and `tex` will now render LaTeX (or rather, the MathJax flavor of it)

Screen Shot 2022-11-27 at 7 54 14 PM

Docs: https://github.com/terrastruct/d2-docs/pull/15

closes #135

### CODE REVIEW

File: d2renderers/d2latex/latex.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	wf, err := strconv.ParseFloat(wEx, 64)
+	if err != nil {
+		return 0, 0, fmt.Errorf("svg parsing failed for latex: %v", svg)
+	}
+	hf, err := strconv.ParseFloat(hEx, 64)
+	if err != nil {
+		return 0, 0, fmt.Errorf("svg parsing failed for latex: %v", svg)
+	}
+
+	return int(math.Ceil(wf * 8)), int(math.Ceil(hf * 8)), nil
```
Comment (ejulio-ts) [inline d2renderers/d2latex/latex.go:78]: Why `* 8` here? 🤔

File: d2renderers/d2latex/latex.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+//go:embed polyfills.js
+var polyfillsJS string
+
+//go:embed setup.js
+var setupJS string
+
+//go:embed mathjax.js
+var mathjaxJS string
+
+var svgRe = regexp.MustCompile(`<svg[^>]+width="([0-9\.]+)ex" height="([0-9\.]+)ex"[^>]+>`)
```
Comment (ejulio-ts) [inline d2renderers/d2latex/latex.go:25]: Maybe just an example that it matches `width=".." height=".."`

Comment (ejulio-ts) [reply d2renderers/d2latex/latex.go:25]: is it true that `width` will always come before `height` and nothing will be in between?

File: d2renderers/d2latex/latex.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	if len(dims) != 1 || len(dims[0]) != 3 {
+		return 0, 0, fmt.Errorf("svg parsing failed for latex: %v", svg)
+	}
+
+	wEx := dims[0][1]
+	hEx := dims[0][2]
+
+	wf, err := strconv.ParseFloat(wEx, 64)
+	if err != nil {
+		return 0, 0, fmt.Errorf("svg parsing failed for latex: %v", svg)
```
Comment (ejulio-ts) [inline d2renderers/d2latex/latex.go:71]: not `sag parsing failed`, `float parsing failed` and then you can use `wEx` and `hEx` or you can get rid of them

File: d2renderers/d2latex/latex_test.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+		if err := xml.Unmarshal([]byte(svg), &xmlParsed); err != nil {
+			t.Fatalf("invalid SVG: %v", err)
+		}
+	}
+}
+
+func TestRenderError(t *testing.T) {
+	_, err := Render(`\frac{1}{2}`)
+	if err == nil {
+		t.Fatal(err)
```
Comment (ejulio-ts) [inline d2renderers/d2latex/latex_test.go:28]: what will it log if `err` is `nil` here? 
also, you can make it an one liner `if _, err := ..`

File: e2etests/stable_test.go [hunk 5]
```diff
@@ -961,6 +961,31 @@ beta: {
 alpha -> beta: gamma {
 	style.font-color: green
 }
+`,
+		},
+		{
+			name: "latex",
+			script: `a: |latex
+\\Huge{\\frac{\\alpha g^2}{\\omega^5} e^{[ -0.74\\bigl\\{\\frac{\\omega U_\\omega 19.5}{g}\\bigr\\}^{\\!-4}\\,]}}
```
Comment (ejulio-ts) [inline e2etests/stable_test.go:969]: why does it need to be `\\`?
I'd expect `\\` to render the actual `\` and not as a prefix to a latex command

**Review by ejulio-ts — CHANGES_REQUESTED** [review]
Edges are handled
```
a: |latex
x = \\frac{1}{2}
|

b: |latex
e = mc^2
|

a -> b: |latex
e = mc^2
|
```

If we're not going to handle them, we should show an error
If we're going to handle it in another PR, totally fine

I'm used to having `latex` inside markdown with `$$` tags `mathjax`, but it doesn't seem we support it right now. I get that t it is probably out of scope for this PR, so just letting you know inc are you want to open an issue

File: e2etests/stable_test.go (hunk 5 above)
Comment (alixander) [reply e2etests/stable_test.go:969]: https://github.com/terrastruct/d2-docs/pull/15/files#diff-d23c28129c5e819c092f20bfcbf291203169d455eabed54a3498779040e06b20R36

File: d2renderers/d2latex/latex_test.go (hunk 4 above)
Comment (alixander) [reply d2renderers/d2latex/latex_test.go:28]: oh right lol

**Comment by ejulio-ts** [comment]
Don't you need to add some `NOTICE` for `mathjax`?

**Comment by alixander** [comment]
>  Don't you need to add some NOTICE for mathjax?

even better -- the attribution is on the file itself (top)

File: d2renderers/d2latex/latex.go (hunk 2 above)
Comment (alixander) [reply d2renderers/d2latex/latex.go:25]: i haven't verified in MathJax's source code, but I see no reason why it'd change

File: d2renderers/d2latex/latex.go (hunk 3 above)
Comment (alixander) [reply d2renderers/d2latex/latex.go:71]: float parsing must've failed because svg parsing failed though, it's a more informative error statement

**Review by ejulio-ts — APPROVED** [review]

File: d2renderers/d2svg/d2svg.go [hunk 6]
```diff
@@ -701,17 +702,27 @@ func drawShape(writer io.Writer, targetShape d2target.Shape) error {
...
 		case d2target.ShapeText:
-			render, err := textmeasure.RenderMarkdown(targetShape.Label)
-			if err != nil {
-				return err
+			if targetShape.Language == "latex" {
+				render, err := d2latex.Render(targetShape.Label)
+				if err != nil {
+					return err
+				}
+				fmt.Fprintf(writer, `<g transform="translate(%f %f)" style="opacity:%f">`, box.TopLeft.X, box.TopLeft.Y, targetShape.Opacity)
```
Comment (gavin-ts) [inline d2renderers/d2svg/d2svg.go:710]:
```suggestion
				fmt.Fprintf(writer, `<g transform="translate(%f, %f)" style="opacity:%f">`, box.TopLeft.X, box.TopLeft.Y, targetShape.Opacity)
```

Comment (alixander) [reply d2renderers/d2svg/d2svg.go:710]: that's not right: https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/transform#translate

**Review by nhooyr — COMMENTED** [review]
nice

**Pull request merged**